### PR TITLE
[REF] l10n_in: fiscal position

### PR DESCRIPTION
**Pull request by khva-odoo**
In this PR
-------------
### [IMP] account: give priority to current company fiscal position
- Before this commit:
If a company is having a branch then while creating a invoice,
fiscal position from parent company is auto applied(auto apply)
even if the branch is having it's own fiscal position with auto
apply true.

- After this commit:
Now if the branch is having it's own fiscal position and
auto apply is True, we prioritize branch fiscal position
over the parent. Which resolves the above mentioned issue

### [REF] l10n_in: refactor fiscal positions

Refactors fiscal position for l10n_in
Removing the loading of fiscal positions from csv
and adding loading of fiscal position from python

Also now if company having new branches, individual
new fiscal position will be created for indian branch

Previously only supported for:
- Inter State
- Reverse Charge Intra State
- Reverse Charge inter State

Now l10n_in supports:
- Intra State (dynamic name based on state name and auto-apply)
- Inter State (auto-apply)
- Outside India (auto-apply)
- SEZ
- LUT SEZ
- LUT Outside India

task id: 3614541

### CODE REVIEW

File: addons/l10n_in/models/company.py [hunk 1]
```diff
@@ -12,3 +11,44 @@ class ResCompany(models.Model):
...
+                company.update_fiscal_position()
+        return res
+
+    def _get_fiscal_vals(self):
+        self.ensure_one()
+        indian_all_states = self.env['res.country.state'].search([('country_id.code', '=', 'IN')])
+        return {
+            'auto_apply': True,
+            'country_id': self.country_id.id,
+            'state_ids': [Command.set((indian_all_states - self.state_id).ids)]
```
Comment (jco-odoo) [inline addons/l10n_in/models/company.py:33]: I think you can avoid this.  Why not use the inherit of the load_data of the CoA to add a fiscal position in auto_apply with a sequence that comes first and the state_id == the state of the company (=within the state, intrastate)) and then the second fiscal position will always just have country India and be called interstate (and we already have it).  @jva-odoo

Comment (ajjo-odoo) [reply addons/l10n_in/models/company.py:33]: Thanks for your input @JCO

The original object of this task was as follows :

upon installation of l10n_in
1. Inter State fiscal position get created automatically ( as working now )
2. addition to above  - Country = company country, state = except company state,
3. Auto detect = True.

Comment (ajjo-odoo) [reply addons/l10n_in/models/company.py:33]: i guess it's the same as you guided by code. we exactly expecting the same,
i guess jva gave it another height that to cover advance cases,
currently we recommend user to manage advance cases manually.

Comment (jco-odoo) [reply addons/l10n_in/models/company.py:33]: Yes, but you do not have the complexity of adding all other states.  You just add the state it is in and you apply the fiscal position before the others.  You can actually quickly configure in a db to try.

File: addons/l10n_in/data/template/account.fiscal.position-in.csv [hunk 2]
```diff
@@ -1,4 +1,5 @@
 "id","name","tax_ids/tax_src_id","tax_ids/tax_dest_id"
```
Comment (jco-odoo) [inline addons/l10n_in/data/template/account.fiscal.position-in.csv:1]: To be explicit, it might not be bad to add the sequence field.

File: addons/l10n_in/data/template/account.fiscal.position-in.csv [hunk 3]
```diff
@@ -11,18 +12,6 @@
 "","","sgst_purchase_12","igst_purchase_12"
 "","","sgst_purchase_18","igst_purchase_18"
 "","","sgst_purchase_28","igst_purchase_28"
-"fiscal_position_in_export","Export","sgst_sale_1","igst_sale_1"
```
Comment (jco-odoo) [inline addons/l10n_in/data/template/account.fiscal.position-in.csv:14]: Should not this be mapped to a 0% instead?

File: addons/l10n_in/models/company.py [hunk 4]
```diff
@@ -12,3 +11,54 @@ class ResCompany(models.Model):
...
+            'fiscal_position_in_intra_state': {
+                'sequence': 1,
+                'auto_apply': True,
+                'country_id': self.country_id.id,
+                'state_ids': [Command.set(self.state_id.ids)]
+            },
+            'fiscal_position_in_inter_state': {
+                'sequence': 2,
+                'auto_apply': True,
+            },
```
Comment (jco-odoo) [inline addons/l10n_in/models/company.py:41]: Why is this needed?
```suggestion
```

File: addons/l10n_in/models/template_in.py [hunk 5]
```diff
@@ -49,3 +49,10 @@ def _get_in_account_cash_rounding(self):
                 'loss_account_id': 'p213201',
             }
         }
+
+    @template('in', 'account.fiscal.position')
+    def _update_fiscal_position(self):
```
Comment (jco-odoo) [inline addons/l10n_in/models/template_in.py:54]: No need to have it in the csv if we have it here.

File: addons/l10n_in/models/company.py [hunk 6]
```diff
@@ -12,3 +11,54 @@ class ResCompany(models.Model):
...
+        }
+
+    def _update_l10n_in_fiscal_position(self):
+        self.ensure_one()
+        for _xml_id, vals in self._get_l10n_in_fiscal_positions_vals_by_ref().items():
```
Comment (jco-odoo) [inline addons/l10n_in/models/company.py:46]: Why so complicated?  We need to just update that one fiscal position with the correct state.  

Of course, the fiscal positions can be shared among branches and that complicates matters...  I do not see how that can work here.  (suppose your parent company is state A and your child company is state B, then state A will still be seen as intrastate).  Another way might be to change the company domain just for this fiscal position... (but that is hacky)

**Review by jco-odoo — COMMENTED** [review]
Trickier than at first, because what do you do in case of branches (which state will rule?)  We could of course just do a checkbox on the fiscal position 'intra-state'.  @tsb-odoo

File: addons/l10n_in/models/company.py (hunk 4 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/company.py:41]: May be we can keep a empty dict? it might be useful for setting up branches (if we may move sequence and auto_apply to csv as you suggested) and easy to update vals in future if required

File: addons/l10n_in/models/template_in.py (hunk 5 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:54]: @jco-odoo can you clarify? are you referring to this https://github.com/odoo/odoo/pull/147812#discussion_r1453145217 ?

File: addons/l10n_in/models/company.py (hunk 6 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/company.py:46]: Well I still need to work on this, it is not going to be an easy job for sure, may be only inheriting `write` method won't be enough

File: addons/l10n_in/data/template/account.fiscal.position-in.csv (hunk 3 above)
Comment (hamo-odoo) [reply addons/l10n_in/data/template/account.fiscal.position-in.csv:14]: @jva-odoo should we do it or not?

**Comment by hamo-odoo** [comment]
>  Another way might be to change the company domain just for this fiscal position... (but that is hacky)

@jco-odoo I have added a company filter for it. Tho a bit hacky but does the job

File: addons/l10n_in/data/template/account.fiscal.position-in.csv (hunk 3 above)
Comment (jva-odoo) [reply addons/l10n_in/data/template/account.fiscal.position-in.csv:14]: No, we need it new tax that effect is 0 like 18% GST but has a repartition line of +100% and -100%
It also depends on Customer holds a Letter of Undertaking(LUT) or not.
so no automation for that

File: addons/l10n_in/models/company.py (hunk 6 above)
Comment (jco-odoo) [reply addons/l10n_in/models/company.py:53]: @hamo-odoo I think there is a way to do it (to avoid the problem if selecting the wrong state fiscal position from a parent branch) if you set the sequence field as well on the fiscal position (child branch with higher priority than parent branch), but we are still thinking if we could do something in the fiscal position auto_apply logic itself.

File: addons/l10n_in/models/template_in.py (hunk 5 above)
Comment (jco-odoo) [reply addons/l10n_in/models/template_in.py:54]: This method should already create it when installing the CoA.  (and putting it in the csv does the same)

Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:54]: Still we might need this method because company already having state while setting up CoA, sets the intrastate state for the Fiscal

Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:54]: So @jco-odoo, are you suggesting that we completely remove the fiscal intrastate and interstate data from csv and we return a complete set of data of intrastate and interstate from here?

That would work thanks for the suggestion :new_moon_with_face:

File: addons/l10n_in/models/company.py [hunk 7]
```diff
@@ -12,3 +11,91 @@ class ResCompany(models.Model):
...
+            elif fiscal_position:
+                fiscal_position.update(vals)
+
+    def _set_l10n_in_branch_fiscal_position(self, data_xml_id, vals):
+        self.ensure_one()
+        parent_fiscal_position = self.env.ref(f"account.{self.parent_id.id}_{data_xml_id}", raise_if_not_found=False)
+        if parent_fiscal_position:
+            vals.update({'company_id': self.id})
+            fiscal_position = parent_fiscal_position.copy(vals)
+            parent_fiscal_position.sequence += 2
```
[the remark this answers is not in the document]
Comment (jco-odoo) [reply addons/l10n_in/models/company.py:96]: @hamo-odoo Are you sure?  The difference in sequence between the 2 parent positions could be greater than 1.

[the remark this answers is not in the document]
Comment (hamo-odoo) [reply addons/l10n_in/models/company.py:96]: Yes, initial we are setting the sequence of intrastate(1) and interstate(2), when there will be new branch the they will copy the sequence of parent and i.e the parent sequence will be 3 and 4(after copy method) while child sequence will be 1 and 2

[the remark this answers is not in the document]
Comment (hamo-odoo) [reply addons/l10n_in/models/company.py:96]: Sometimes it really gets me there, there are few issues that really needs fixed/improved :thinking: or as of alternative we can go with the filter if you agree with that?
```xml
<?xml version="1.0" encoding="UTF-8"?>
<odoo>
    <record id="account_fiscal_position_search_inherit_l10n_in" model="ir.ui.view">
        <field name="name">account.fiscal.position.search.inherit.l10n_in</field>
        <field name="model">account.fiscal.position</field>
        <field name="inherit_id" ref="account.view_account_position_filter"/>
        <field name="arch" type="xml">
            <xpath expr="//filter[@name='active']" position="after">
                <filter string="Company" name="by_company" domain="[('company_id', 'in', allowed_company_ids)]"/>
            </xpath>
        </field>
    </record>
    <record id="account.action_account_fiscal_position_form" model="ir.actions.act_window">
        <field name="context">{'search_default_by_company': 1}</field>
    </record>
</odoo>
```

**Comment by hamo-odoo** [comment]
Updated the PR as per the updated spec of task

**Comment by hamo-odoo** [comment]
To make sure companies and branches intrastate are mapped properly I have added test cases for it :see_no_evil:

File: addons/l10n_in/models/company.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hamo-odoo) [reply addons/l10n_in/models/company.py]: @jco-odoo Do we add filter or not? Because now we got 6 fiscal positions (as per the final specs decided)

File: addons/l10n_in/models/company.py [hunk 8]
```diff
@@ -12,3 +11,130 @@ class ResCompany(models.Model):
...
+            'fiscal_position_in_outside_in': self._get_fiscal_position_in_outside_in_vals(load_template),
+            'fiscal_position_in_lut_sez': self._get_fiscal_position_in_lut_sez_vals(load_template),
+            'fiscal_position_in_lut_outside_in': self._get_fiscal_position_in_lut_outside_in_vals(load_template),
+        }
+
+    def _update_l10n_in_fiscal_position(self):
+        self.ensure_one()
+        for xml_id, vals in self._get_l10n_in_fiscal_positions_vals_by_ref().items():
+            fiscal_position = self.env['account.chart.template'].ref(xml_id, raise_if_not_found=False)
+            if fiscal_position and fiscal_position.company_id == self.parent_id and self.parent_id.account_fiscal_country_id.code == 'IN':
```
Comment (jva-odoo) [inline addons/l10n_in/models/company.py:126]:
```suggestion
            main_company = self.parent_id or self
            fiscal_position = self.env['account.chart.template'].with_company(main_company).ref(xml_id, raise_if_not_found=False)
            if fiscal_position and fiscal_position.company_id == self.parent_id and self.parent_id.account_fiscal_country_id.code == 'IN':
```

**Comment by hamo-odoo** [comment]
@jco-odoo Can you please re-review this PR? I guess I had make some minor changes in `account` module for this task

File: addons/l10n_in/models/company.py [hunk 9]
```diff
@@ -12,3 +11,131 @@ class ResCompany(models.Model):
...
+        return {}
+    def _get_fiscal_position_in_lut_outside_in_vals(self, load_template):
+        if load_template:
+            return {
+                'name': "LUT Outside India",
+                'sequence': 6,
+                'tax_ids': self._get_l10n_in_fiscal_tax_vals(use_zero_rated_igst=True),
+            }
+        return {}
+    def _get_l10n_in_fiscal_positions_vals_by_ref(self, load_template=False):
```
Comment (william-andre) [inline addons/l10n_in/models/company.py:112]: What is the use of this `load_template` param? What is the effect when it's false?

**Review by william-andre — COMMENTED** [review]
> This commits refactors fiscal position for l10n_in Removing the loading of fiscal positions from csv and adding loading of fiscal position from python

Why?

**Review by Levizar — CHANGES_REQUESTED** [review]
A first review for the first commit

File: addons/l10n_in/models/company.py (hunk 9 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/company.py:112]: `load_template` will be only used (True) when we are loading the templates from `template_in.py` in other cases we require to create fiscals for the branches or update it when you set the state for the first time

**Comment by hamo-odoo** [comment]
> Why?

@william-andre @tsb-odoo, 
We decided to refactor it because to improve on boarding whenever a user will select company state_id for the first time we want a fiscal for intra state (within that state to be applied on the invoice i.e. GST should be applied) similarly doing it for other fiscals too, in case if the customer didn't select the state before COA loaded it would be the same data we want to override fiscal data whenever user sets the state of the company for the first time (similar use case we are doing it for autoinstall of `l10n` modules), Hence we are already preparing vals in py then why have duplicate data in csv too.

File: addons/account/models/partner.py [hunk 10]
```diff
@@ -213,12 +213,19 @@ def _get_vat_valid(self, delivery, company=None):
...
-            *self._check_company_domain(self.env.company),
+        domain = [
             ('auto_apply', '=', True),
             ('vat_required', '=', vat_required),
         ]
+        if self._check_current_company_have_auto_apply():
+            domain.append(('company_id', '=', self.env.company.id))
+        else:
+            domain.extend(self._check_company_domain(self.env.company))
+        return domain
```
[the remark this answers is not in the document]
Comment (hamo-odoo) [reply addons/account/models/partner.py:228]: @Levizar Nice catch and Thanks for the suggestions!
Can you please explain the case when it goes to fallback (with no state and no zip)? if that is the case shouldn't it be the first domain search it would be? Still thinking about the cases where we can still optimize it :thinking:

**Comment by william-andre** [comment]
@khva-odoo @hamo-odoo This is probably too naive, but please review this new way of doing things

**Comment by hamo-odoo** [comment]
> @khva-odoo @hamo-odoo This is probably too naive, but please review this new way of doing things

Thanks @william-andre for the fixup, We will surely take this things into consideration couldn't ask for any better :bowing_man:

**Comment by hamo-odoo** [comment]
image
@william-andre This view looks abit confusing for the user (though we have the company on the right side) when the active company is a branch

As for a suggestion I do have a filter for it https://github.com/odoo/odoo/pull/147812#discussion_r1464663792
What do you think about it?

File: addons/l10n_in/models/template_in.py [hunk 11]
```diff
@@ -47,3 +48,61 @@ def _get_in_account_cash_rounding(self):
                 'loss_account_id': 'p213201',
             }
         }
+
+    @template('in', 'account.fiscal.position')
+    def _get_in_account_fiscal_position(self):
+        company = self.env.company
+        state_ids = [Command.set(company.state_id.ids)] if company.state_id else False
+        intra_state_name = company.state_id and 'Within %s' % company.state_id.name or 'Intra State'
```
Comment (william-andre) [inline addons/l10n_in/models/template_in.py:56]: This is missing translations (check the whole diff/file)

**Review by william-andre — COMMENTED** [review]
I added 2 fixups:
* One in `account` to avoid doing many `search` growing exponentially... I also find the implementation easier to understand and to extend. It also allows for a more complex hierarchy of branches
* On in `l10n_in` to avoid creating useless fiscal positions (when they do exactly the same thing between the branch and the main company)

File: addons/account/models/partner.py [hunk 12]
```diff
@@ -172,52 +172,34 @@ def write(self, vals):
...
+            )),
+            ('country_id', lambda fpos: (
+                not fpos.country_id
+                or (partner.country_id == fpos.country_id and 2)
+            )),
+            ('country_group', lambda fpos: (
+                not fpos.country_group_id
+                or (partner.country_id in fpos.country_group_id.country_ids and 2)
+            )),
+            ('sequence', lambda fpos: fpos.sequence),
```
Comment (william-andre) [inline addons/account/models/partner.py:202]: Tests are probably failing because there is no default value, so we need it to not be falsy
```suggestion
            ('sequence', lambda fpos: fpos.sequence or 0.1),  # do not filter out sequence=0
```

**Review by william-andre — COMMENTED** [review]
To fix the tests

**Comment by hamo-odoo** [comment]
@william-andre May be @jco-odoo or @Levizar can do the final review and sign it off as they have already done the review initially

File: addons/account/models/partner.py [hunk 13]
```diff
@@ -248,15 +230,22 @@ def _get_fiscal_position(self, partner, delivery=None):
...
+        def ranking_function(fpos):
+            return tuple(rank[1](fpos) for rank in ranking_subfunctions)
+
+        return max(
+            filter(lambda x: all(x[1]), [
+                (fpos, ranking_function(fpos))
+                for fpos in self.search([
+                    *self._check_company_domain(self.env.company),
+                    ('auto_apply', '=', True),
+                ])
```
Comment (william-andre) [inline addons/account/models/partner.py:244]: If we wanted to get almost completely rid of any `search` to find fiscal positions in loops, we could store (in memory) the fiscal positions of a company on a non stored (in db) field of the company.

Probably over engineering for now though.

**Review by william-andre — APPROVED** [review]
Yes, as I wrote most of the code (outside of the tests), I won't be signing this off.
But it is ready to be reviewed according to me. :+1:

File: addons/account/models/partner.py [hunk 14]
```diff
@@ -172,52 +172,34 @@ def write(self, vals):
...
 
-    def _prepare_fpos_base_domain(self, vat_required):
+    def _get_fpos_ranking_functions(self, partner):
         return [
-            *self._check_company_domain(self.env.company),
-            ('auto_apply', '=', True),
-            ('vat_required', '=', vat_required),
+            ('vat_required', lambda fpos: (
+                not fpos.vat_required
+                or (self._get_vat_valid(partner, self.env.company) and 2)
```
Comment (jco-odoo) [inline addons/account/models/partner.py:183]: Something is off: This is only 0 when the vat is required on the fpos and there is no vat, while one or the other way, the logic was changed before and vat_required has to be equal to vat_required.  I could say: put it back  in the base domain, but this might work as well:  (just to follow the same logic in a very special case where zip would be set/not set e.g.)
```suggestion
                or (self._get_vat_valid(partner, self.env.company) and 200)
```

But this does not work when there is no vat on the partner, because then it should never get into the vat required ones.  So we need (not fpos.required and not self._get_vat_valid(partner, company) and 1 or 0)  (but then the terms need to be juggled)

So we need something like:
self._get_vat_valid(partner, self.env.company) and 200 or 1 if fpos.vat_required else not self._get_vat_valid(partner, company)

File: addons/account/models/partner.py [hunk 15]
```diff
@@ -172,52 +172,34 @@ def write(self, vals):
...
+            )),
+            ('company_id', lambda fpos: len(fpos.company_id.parent_ids)),
```
Comment (jco-odoo) [inline addons/account/models/partner.py:185]: Luckily, I am a parent of myself...

**Review by jco-odoo — COMMENTED** [review]
Fiscal position logic is complicated.  

I wish the functional approach could be a bit more explicit all(x[1]  if we just put a comment like: all values need to be true or we do the score maybe in the ranking method.  Also that method does not necessary need to return a function and why do we bother which key has which score. 

Otherwise, in terms of the Indian multi-branches, it seems to make it a lot simpler.  It means however that fiscal positions with the same number of other conditions will have priority in the child branch, but they could in theory still switch between parent and child branches.  I wonder e.g. if when we set the states/zip codes we also set the country and as such it can precede over the one with only country.  Maybe the geographical thing should be just one score.  It should be clear somehow how parent and child branches intermingle.  (not sure if we should give the parent_ids a higher score as well)

**Review by william-andre — COMMENTED** [review]
> (not sure if we should give the parent_ids a higher score as well)

This is the best approach I can think of regarding the branches: using the `len(company.parent_ids)` ensures that it will always take the first compatible fpos working up the tree of company branches one level at a time

File: addons/account/models/partner.py [hunk 16]
```diff
@@ -248,15 +230,22 @@ def _get_fiscal_position(self, partner, delivery=None):
...
+            filter(lambda x: all(x[1]), [
+                (fpos, ranking_function(fpos))
+                for fpos in self.search([
+                    *self._check_company_domain(self.env.company),
+                    ('auto_apply', '=', True),
+                ])
+            ]),
+            key=lambda x: x[1],
+            default=(self.env['account.fiscal.position'], False)
+        )[0]
```
Comment (Levizar) [inline addons/account/models/partner.py:248]: Several stuff:
- if there isn't any fiscal position satisfying every criteria, it won't be selected (because of the `all(critera_tuple)`)
At first sight, it's nice because it means you will get the perfect one but what if there isn't any that correspond but one is nearly the good one? Do we want to not set any because it's not close enough?
- This looks cleaver but is very hard to read. It would be better to use intermediary variables like `all_auto_apply_fpos` for the search, `ranked_fpos` for the list comp, `eligible_fpos` for the filter.
- nitpick: list comp could be a gen comp, no need to materialize the list

File: addons/account/models/partner.py [hunk 17]
```diff
@@ -172,52 +172,34 @@ def write(self, vals):
...
-            # Fallback on catchall (no country, no group)
-            fpos = self.search(base_domain + null_country_dom, limit=1)
-        return fpos
-
     def _get_vat_valid(self, delivery, company=None):
         """ Hook for determining VAT validity with more complex VAT requirements """
         return bool(delivery.vat)
 
-    def _prepare_fpos_base_domain(self, vat_required):
+    def _get_fpos_ranking_functions(self, partner):
```
Comment (Levizar) [inline addons/account/models/partner.py:179]: We should add a comment explaining how it works.
I didn't figured directly that, for example, `vat_required` not being set means that the criterion is thus already satisfied.

File: addons/account/models/partner.py (hunk 14 above)
Comment (Levizar) [reply addons/account/models/partner.py:183]: I didn't get it either @jco-odoo :/

Comment (william-andre) [reply addons/account/models/partner.py:183]: Do you know how tuple comparison works? :smile:

Comment (jco-odoo) [reply addons/account/models/partner.py:183]: So, tuple comparison, no sum.

File: addons/l10n_in/models/template_in.py [hunk 18]
```diff
@@ -47,3 +48,61 @@ def _get_in_account_cash_rounding(self):
...
+            'fiscal_position_in_lut_sez': {
+                'name': _('LUT - Special Economic Zone (SEZ)'),
+                'sequence': 5,
+                'tax_ids': self._get_l10n_in_fiscal_tax_vals(use_zero_rated_igst=True),
+            },
+            'fiscal_position_in_lut_outside_in': {
+                'name': _("LUT Outside India"),
+                'sequence': 6,
+                'tax_ids': self._get_l10n_in_fiscal_tax_vals(use_zero_rated_igst=True),
+            },
```
Comment (william-andre) [inline addons/l10n_in/models/template_in.py:97]: What's the difference between those 2 fiscal positions?

File: addons/l10n_in/models/template_in.py [hunk 19]
```diff
@@ -47,3 +48,61 @@ def _get_in_account_cash_rounding(self):
...
+                'name': _('Outside India'),
+                'sequence': 3,
+                'auto_apply': True,
+                'tax_ids': self._get_l10n_in_fiscal_tax_vals(),
+            },
+            'fiscal_position_in_sez': {
+                'name': _('Special Economic Zone (SEZ)'),
+                'sequence': 4,
+                'tax_ids': self._get_l10n_in_fiscal_tax_vals(),
+            },
```
Comment (william-andre) [inline addons/l10n_in/models/template_in.py:87]: What's the difference between those 2 fiscal positions? (except the auto apply)

File: addons/l10n_in/data/template/account.fiscal.position-in.csv [hunk 20]
```diff
@@ -1,28 +1,4 @@
...
-"","","sgst_sale_12","igst_sale_12"
-"","","sgst_sale_18","igst_sale_18"
-"","","sgst_sale_28","igst_sale_28"
-"","","sgst_purchase_1","igst_purchase_1"
-"","","sgst_purchase_2","igst_purchase_2"
-"","","sgst_purchase_5","igst_purchase_5"
-"","","sgst_purchase_12","igst_purchase_12"
-"","","sgst_purchase_18","igst_purchase_18"
-"","","sgst_purchase_28","igst_purchase_28"
 "fiscal_position_in_reverse_charge_intra","Reverse charge Intra State","sgst_purchase_1","sgst_purchase_1_rc"
```
Comment (william-andre) [inline addons/l10n_in/data/template/account.fiscal.position-in.csv:2]: I suppose this is never auto applied?

File: addons/l10n_in/models/template_in.py (hunk 19 above)
Comment (jco-odoo) [reply addons/l10n_in/models/template_in.py:87]: I suppose it is difficult to have a criterium in auto_apply for the Special Economic Zone cases.  @hamo-odoo

File: addons/l10n_in/models/template_in.py (hunk 18 above)
Comment (jco-odoo) [reply addons/l10n_in/models/template_in.py:97]: @ajjo-odoo @jva-odoo @hamo-odoo

File: addons/l10n_in/models/template_in.py (hunk 19 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:87]: About Outside India:
Transaction occurring with foreign entity

About SEZ:
A Special Economic Zone (SEZ) is a geographical region that has economic laws that are more liberal than a country's domestic economic laws.

SEZs in India are areas that offer incentives to resident businesses. SEZs typically offer competitive infrastructure, duty free exports, tax incentives, and other measures designed to make it easier to conduct business. Accordingly, SEZs in India are a popular investment destination for many multinationals, particularly exporters.

cc- @ajjo-odoo

File: addons/l10n_in/models/template_in.py (hunk 18 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:97]: About **LUT Outside India** :
 Letter of Undertaking holds significant relevance within the context of the Goods and Services Tax (GST) framework. This document serves as a powerful tool for exporters, allowing them to engage in the export of goods or services without the obligation of immediate tax payment

About **LUT - SEZ** :
SEZ units can make inward supplies in two ways as follows:

- Without payment of IGST (With LUT) 
 - With payment of IGST (Without LUT) and claiming Input Tax Credit

The suppliers of SEZ unit/developer are required to issue LUT so they don’t need to collect GST.  

However, if they don’t issue LUT, they have to collect IGST and then SEZ unit/developer can avail refund of such Input Tax Credit under GST

cc- @ajjo-odoo

File: addons/l10n_in/models/template_in.py (hunk 19 above)
Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:87]: > I suppose it is difficult to have a criterium in auto_apply for the Special Economic Zone cases. @hamo-odoo

We don't have any plan to do it as of now (we will only end up making things complex)

**Review by hamo-odoo — COMMENTED** [review]
I hope it clears your doubts regarding the fiscal

Feel free to correct me @ajjo-odoo not that knowledgeable as a CA but still learning :smiley:

File: addons/l10n_in/models/template_in.py (hunk 18 above)
Comment (william-andre) [reply addons/l10n_in/models/template_in.py:97]: Ok, but why do we need 2 separate fiscal positions?
Does it make sense? Will they ever be different?

Comment (hamo-odoo) [reply addons/l10n_in/models/template_in.py:97]: > Ok, but why do we need 2 separate fiscal positions?

For Functional purposes both will be treated different for reporting

> Does it make sense? Will they ever be different?

Technically not but yes for functional reporting of both fiscal will be different

Comment (jco-odoo) [reply addons/l10n_in/models/template_in.py:97]: It is preferable not to use the use of fiscal position for that, because you can always put the taxes manually without the fiscal position.  Should not we have separate taxes for it if they are reported differently?  @ajjo-odoo

Comment (ajjo-odoo) [reply addons/l10n_in/models/template_in.py:97]: 1) Within Gujarat ( company state )    (Intra State if not configured; rename when company state changes) (No tax mapping)
==> If supply made between same state ( company state = partner state ) Exception if SEZ.

2) Inter-state (GST -> IGST)
==> Company State is diffrent than Partner State ( In case of India ) 

3) Special Economic Zone (SEZ) (GST -> IGST)
==> State could be same or different - Government declare specific area as a SEZ where in supply
as fiscal position does not support parameter of GST Treatment type to auto apply so, it set for manuall selection.

4) Outside India without LUT (GST -> IGST)
==> Country is not equal to India - here supplier will pay a Tax to Government on supply of goods and will get credit back

5) Outside India with LUT (GST -> 0% IGST)

==> Country is not equal to India - here supplier will not pay a tax as he will hold a certificate of LUT and tax rate would be Zero percent.

As LUT is based on period and certain conditions so posisbility of supply with LUT & Without for the same company.
So it would be good to keep both.

Comment (jco-odoo) [reply addons/l10n_in/models/template_in.py:97]: @ajjo-odoo What about just one fiscal position "LUT"?

**Comment by jco-odoo** [comment]
@hamo-odoo The plan is to remove 2 fiscal positions by merging Outside India and SEZ.  Please check with @jva-odoo

**Comment by hamo-odoo** [comment]
> @hamo-odoo The plan is to remove 2 fiscal positions by merging Outside India and SEZ. Please check with @jva-odoo

I have made the changes accordingly.

@ajjo-odoo can you please the verify the changes made :sweat_smile: ?

File: addons/l10n_in/models/template_in.py (hunk 18 above)
Comment (ajjo-odoo) [reply addons/l10n_in/models/template_in.py]: > @ajjo-odoo What about just one fiscal position "LUT"?

yeah thanks for the idea, it's fine to have as per our last call.

**Comment by jco-odoo** [comment]
@hamo-odoo Could you do a rebase with master here, so we can move forward?

File: addons/point_of_sale/static/src/app/models/utils/tax_utils.js [hunk 21]
```diff
@@ -57,14 +57,19 @@ export const getTaxesAfterFiscalPosition = (taxes, fiscalPosition, models) => {
...
-        if (fiscalPosition.tax_mapping_by_ids[tax.id]) {
-            for (const mapTaxId of fiscalPosition.tax_mapping_by_ids[tax.id]) {
-                newTaxIds.push(mapTaxId);
+        const fiscalTaxLines = fiscalPosition.tax_ids.filter(
+            (fiscalTax) => fiscalTax.tax_src_id.id == tax.id
+        );
+        if (fiscalTaxLines.length > 0) {
+            for (const line of fiscalTaxLines) {
+                if (line.tax_dest_id) {
+                    newTaxIds(line.tax_dest_id.id);
```
Comment (hamo-odoo) [inline addons/point_of_sale/static/src/app/models/utils/tax_utils.js:66]:
```suggestion
                    newTaxIds.push(line.tax_dest_id.id);
```
Runbot again failed :face_exhaling:

Comment (jco-odoo) [reply addons/point_of_sale/static/src/app/models/utils/tax_utils.js]: @hamo-odoo Did the test fail on this?

**Comment by jco-odoo** [comment]
@william-andre The failing l10n tests in l10n_nl_reports tell me that the fiscal position of the invoice is different because of the partner having a country_id False.  It would find non-intra EU with this PR instead of nothing.  (for invoice_partner_a in the XAF test)

**Comment by jco-odoo** [comment]
@robodoo r+ rebase-ff
